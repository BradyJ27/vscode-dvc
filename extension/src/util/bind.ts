export const bindMethods = <T>(self: T, ...methodNames: string[]) =>
  methodNames.forEach(method => {
    if (typeof self[method as keyof T] === 'function') {
      self[method as keyof T] = ((self[
        method as keyof T
      ] as unknown) as Function).bind(self)
    }
  })
