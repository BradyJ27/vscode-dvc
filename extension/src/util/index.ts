export { execPromise } from './exec'

export const isStringInEnum = (s: string, E: Record<string, string>) =>
  Object.values(E).includes(s)

export const definedAndNonEmpty = (
  maybeArray: unknown[] | undefined
): maybeArray is unknown[] => {
  return !!maybeArray?.length
}
