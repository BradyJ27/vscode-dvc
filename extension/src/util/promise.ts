import { delay } from './time'
import { Logger } from '../common/logger'

const resolve = (promiseOrPromises: Promise<unknown>[] | Promise<unknown>) => {
  if (Array.isArray(promiseOrPromises)) {
    return Promise.all(promiseOrPromises)
  }
  return promiseOrPromises
}

export const retryUntilAllResolved = async <T>(
  getNewPromiseOrPromises: () => Promise<unknown>[] | Promise<unknown>,
  type: string,
  waitBeforeRetry = 500
): Promise<T> => {
  try {
    const promiseOrPromises = getNewPromiseOrPromises()
    const data = await resolve(promiseOrPromises)
    return (data as unknown) as T
  } catch (e) {
    Logger.error(`${type} failed with ${e} retrying...`)
    await delay(waitBeforeRetry)
    return retryUntilAllResolved(
      getNewPromiseOrPromises,
      type,
      waitBeforeRetry * 2
    )
  }
}
