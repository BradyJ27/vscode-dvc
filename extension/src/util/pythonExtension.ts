import { extensions, Extension } from 'vscode'

export interface PythonExtensionAPI {
  ready: Thenable<void>
  settings: {
    getExecutionDetails: () => {
      execCommand: string[] | undefined
    }
  }
}

export type PythonExtension = Extension<PythonExtensionAPI>

export const getPythonExtension: () => Thenable<
  PythonExtension | undefined
> = async () => {
  const extension = extensions.getExtension('ms-python.python')
  return extension || undefined
}

export const getActivatedPythonExtension: () => Thenable<
  PythonExtension | undefined
> = async () => {
  const extension = await getPythonExtension()
  if (!extension) return extension
  if (!extension.isActive) await extension.activate()
  return extension
}

export const getReadyPythonExtension: () => Thenable<
  PythonExtension | undefined
> = async () => {
  const extension = await getActivatedPythonExtension()
  if (!extension) return extension
  await extension.exports.ready
  return extension
}

export const getPythonExecutionDetails: () => Thenable<string[]> = async () => {
  const extension = await getReadyPythonExtension()
  if (!extension) return []
  const { execCommand } = extension.exports.settings.getExecutionDetails()
  return execCommand || []
}
