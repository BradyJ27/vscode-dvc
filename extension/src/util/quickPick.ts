import { QuickPickOptions, QuickPickItem, window } from 'vscode'

export interface QuickPickItemWithValue extends QuickPickItem {
  value: string
}

export const quickPickValue: (
  items: QuickPickItemWithValue[],
  options: Omit<QuickPickOptions, 'canPickMany'>
) => Thenable<string | undefined> = async (items, options) =>
  (await window.showQuickPick<QuickPickItemWithValue>(items, options))?.value

export const quickPickManyValues: (
  items: QuickPickItemWithValue[],
  options: Omit<QuickPickOptions, 'canPickMany'>
) => Thenable<string[] | undefined> = async (items, options = {}) =>
  (
    await window.showQuickPick<QuickPickItemWithValue>(items, {
      ...options,
      canPickMany: true
    })
  )?.map(item => item.value)
