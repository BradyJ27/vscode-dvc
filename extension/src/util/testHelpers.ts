import { Uri } from 'vscode'

export function mapPaths(uris?: Uri[]): string[] | undefined {
  return uris?.map(uri => uri.fsPath)
}
