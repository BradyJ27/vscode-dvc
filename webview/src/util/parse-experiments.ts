import {
  ExperimentsRepoJSONOutput,
  ExperimentFields
} from 'dvc/src/experiments/contract'

export interface ExperimentWithSubRows extends ExperimentFields {
  id: string
  subRows?: ExperimentWithSubRows[]
}

export type RepoExperiments = Array<ExperimentWithSubRows>

interface ParseExperimentsOutput {
  experiments: ExperimentWithSubRows[]
  flatExperiments: ExperimentWithSubRows[]
}

const addIdToExperiment: (
  id: string,
  experiment: ExperimentFields
) => ExperimentWithSubRows = (id, experiment) => ({
  ...experiment,
  id
})

const buildExperimentFromEntry: (
  entry: [string, ExperimentFields]
) => ExperimentWithSubRows = ([id, experiment]) =>
  addIdToExperiment(id, experiment)

const isTopLevelExperiment: (row: ExperimentWithSubRows) => boolean = ({
  queued,
  id,
  checkpoint_tip
}) => queued || (checkpoint_tip ? id === checkpoint_tip : true)

const pushResult = (
  currentTip: ExperimentWithSubRows | undefined,
  currentEpochs: ExperimentWithSubRows[],
  result: ExperimentWithSubRows[]
): void => {
  if (currentTip) {
    const resultToPush = {
      ...currentTip
    }
    if (currentEpochs.length > 0) {
      resultToPush.subRows = currentEpochs
    }
    result.push(resultToPush)
  }
}

const groupCheckpoints = (
  rows: ExperimentWithSubRows[]
): ExperimentWithSubRows[] => {
  let currentTip: ExperimentWithSubRows | undefined
  let currentEpochs: ExperimentWithSubRows[] = []
  const result: ExperimentWithSubRows[] = []

  for (const row of rows) {
    if (isTopLevelExperiment(row)) {
      pushResult(currentTip, currentEpochs, result)
      currentTip = row
      currentEpochs = []
    } else {
      currentEpochs.push({ ...row })
    }
  }
  pushResult(currentTip, currentEpochs, result)
  return result
}

const parseExperiments: (
  experimentsData: ExperimentsRepoJSONOutput
) => ParseExperimentsOutput = experimentsData => {
  return Object.entries(experimentsData).reduce<ParseExperimentsOutput>(
    (
      { experiments, flatExperiments },
      [commitId, { baseline, ...childExperiments }]
    ) => {
      const parsedChildExperiments: ExperimentWithSubRows[] = Object.entries(
        childExperiments
      ).map<ExperimentWithSubRows>(buildExperimentFromEntry)
      const baselineEntry = addIdToExperiment(commitId, baseline)
      return {
        experiments: [
          ...experiments,
          {
            ...baselineEntry,
            subRows: groupCheckpoints(parsedChildExperiments)
          }
        ],
        flatExperiments: [
          ...flatExperiments,
          baselineEntry,
          ...parsedChildExperiments
        ]
      }
    },
    {
      experiments: [],
      flatExperiments: []
    }
  )
}

export default parseExperiments
