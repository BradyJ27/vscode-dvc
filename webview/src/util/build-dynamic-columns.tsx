/* eslint-disable @typescript-eslint/no-use-before-define */
import React from 'react'
import get from 'lodash/get'

import { Column, Accessor } from 'react-table'
import { ColumnData } from 'dvc/src/experiments/webview/contract'

import { ExperimentWithSubRows } from './parse-experiments'
import { formatFloat } from './number-formatting'

type Value = string | number

const UndefinedCell = <>-</>

const Cell: React.FC<{ value: Value }> = ({ value }) => {
  if (value === undefined) {
    return UndefinedCell
  }

  if (typeof value === 'number' && !Number.isInteger(value)) {
    return <>{formatFloat(value as number)}</>
  }
  return <>{String(value)}</>
}

const getCellComponent = (): React.FC<{ value: Value }> => {
  return Cell
}

const buildColumnIdFromPath = (objectPath: string[]) =>
  objectPath.map(segment => `[${segment}]`).join('')

const buildAccessor: (valuePath: string[]) => Accessor<ExperimentWithSubRows> =
  pathArray => originalRow =>
    get(originalRow, pathArray)

const buildColumnsFromSchemaProperties = (
  properties: ColumnData[],
  objectPath: string[] = []
): Column<ExperimentWithSubRows>[] => {
  return properties.map(data => {
    const currentPath = [...objectPath, data.name]
    const Cell = getCellComponent()
    const column: Column<ExperimentWithSubRows> & {
      columns?: Column<ExperimentWithSubRows>[]
      sortType?: string
      type?: string[]
    } = {
      Cell,
      Header: data.name,
      accessor: buildAccessor(currentPath),
      columns: data?.childColumns?.length
        ? buildColumnsFromSchemaProperties(data.childColumns, currentPath)
        : undefined,
      id: buildColumnIdFromPath(currentPath),
      type: data.types
    }
    switch (data.types) {
      case ['integer']:
      case ['number']:
        column.sortType = 'basic'
        break
      default:
    }
    return column
  })
}

const buildDynamicColumnsFromExperiments = (
  params: ColumnData[],
  metrics: ColumnData[]
): Column<ExperimentWithSubRows>[] => {
  return [
    ...buildColumnsFromSchemaProperties(params, ['params']),
    ...buildColumnsFromSchemaProperties(metrics, ['metrics'])
  ]
}

export default buildDynamicColumnsFromExperiments
