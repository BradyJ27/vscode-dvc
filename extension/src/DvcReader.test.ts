import { getTableData, inferDefaultOptions } from "./DvcReader";
import * as path from "path";
import * as util from "util";

const testDvcDirectory = path.resolve("..", "demo", "example-get-started");
const extensionDirectory = path.resolve();

const testOptions = inferDefaultOptions(testDvcDirectory);

test("Options on the base non-.env repo defaults to dvc on PATH", async () => {
	expect(await inferDefaultOptions(path.resolve())).toEqual({
		bin: "dvc",
		cwd: extensionDirectory,
	});
});

test("Options are properly inferred from the test repo", async () => {
	expect(await testOptions).toEqual({
		bin: path.resolve(testDvcDirectory, ".env", "bin", "dvc"),
		cwd: testDvcDirectory,
	});
});

test("Comparing a table in the test repo to a manual copy of what it should be", async () => {
	const tableData = await getTableData(await testOptions);
	expect(tableData).toEqual({
		workspace: {
			baseline: {
				timestamp: null,
				params: {
					"params.yaml": {
						prepare: { split: 0.2, seed: 20170428 },
						featurize: { max_features: 1500, ngrams: 2 },
						train: { seed: 20170428, n_estimators: 50 },
					},
				},
				queued: false,
				metrics: { "scores.json": { auc: 0.6131382960762474 } },
			},
		},
		"9e05ab4590fb1a569875153bbb8a8a0231e2acec": {
			baseline: {
				timestamp: "2020-08-01T23:01:45",
				params: {
					"params.yaml": {
						prepare: { split: 0.2, seed: 20170428 },
						featurize: { max_features: 1500, ngrams: 2 },
						train: { seed: 20170428, n_estimators: 50 },
					},
				},
				queued: false,
				metrics: { "scores.json": { auc: 0.6131382960762474 } },
				name: "10-bigrams-experiment",
			},
			ddd585e77fdbb49f94452bbb7f2c5d86169d1056: {
				timestamp: "2020-10-19T19:29:26",
				params: {
					"params.yaml": {
						prepare: { split: 0.2, seed: 20170428 },
						featurize: { max_features: 1500, ngrams: 2 },
						train: { seed: 20170428, n_estimators: 50 },
					},
				},
				queued: false,
				metrics: { "scores.json": { auc: 0.6131382960762474 } },
			},
		},
	});
});
