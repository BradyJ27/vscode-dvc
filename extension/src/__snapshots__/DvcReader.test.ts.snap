// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comparing a table in the test repo to a snapshot 1`] = `
Object {
  "experiments": Array [
    Object {
      "baseline": Object {
        "metrics": Object {
          "scores.json": Object {
            "auc": 0.6131382960762474,
          },
        },
        "name": "master",
        "params": Object {
          "params.yaml": Object {
            "featurize": Object {
              "max_features": 1500,
              "ngrams": 2,
            },
            "prepare": Object {
              "seed": 20170428,
              "split": 0.2,
            },
            "train": Object {
              "n_estimators": 50,
              "seed": 20170428,
            },
          },
        },
        "queued": false,
        "timestamp": "2020-10-20T14:52:40",
      },
      "label": "430c6a15611f654ba38d79f435edc4cebf337305",
      "sha": "430c6a15611f654ba38d79f435edc4cebf337305",
    },
  ],
  "workspace": Object {
    "baseline": Object {
      "metrics": Object {
        "scores.json": Object {
          "auc": 0.6131382960762474,
        },
      },
      "params": Object {
        "params.yaml": Object {
          "featurize": Object {
            "max_features": 1500,
            "ngrams": 2,
          },
          "prepare": Object {
            "seed": 20170428,
            "split": 0.2,
          },
          "train": Object {
            "n_estimators": 50,
            "seed": 20170428,
          },
        },
      },
      "queued": false,
      "timestamp": null,
    },
    "label": "workspace",
    "sha": "workspace",
  },
}
`;
